File: tools/grab_mcsm_docx.py
import asyncio
import os
import re
import time
from datetime import datetime
from io import BytesIO
from urllib.parse import urljoin, urlparse

# --- docx ---
from docx import Document
from docx.shared import Inches

# --- http ---
import requests

# --- playwright ---
from playwright.async_api import async_playwright


def safe_filename(name: str, max_len: int = 120) -> str:
    name = (name or "").strip()
    if not name:
        return "output"
    name = re.sub(r'[<>:"/\\|?*\x00-\x1F]', "_", name)
    name = re.sub(r"\s+", " ", name).strip()
    if len(name) > max_len:
        name = name[:max_len].strip()
    return name or "output"


def sanitize_text(s: str) -> str:
    if s is None:
        return ""
    # 清掉 XML 1.0 不允許控制字元，避免 Word 進修復流程
    out_chars = []
    for ch in str(s):
        code = ord(ch)
        if code in (0x09, 0x0A, 0x0D):  # tab, lf, cr
            out_chars.append(ch)
            continue
        if code < 0x20:
            continue
        if 0xD800 <= code <= 0xDFFF:
            continue
        if code in (0xFFFE, 0xFFFF):
            continue
        out_chars.append(ch)

    out = "".join(out_chars)
    out = out.replace("\u00a0", " ")
    out = re.sub(r"\s+", " ", out).strip()
    return out


def uniq_keep_order(items):
    seen = set()
    out = []
    for x in items:
        x = (x or "").strip()
        if not x:
            continue
        if x in seen:
            continue
        seen.add(x)
        out.append(x)
    return out


def normalize_paragraphs(lines):
    # 去掉常見雜訊（你可自行加）
    junk_contains = ["MULTISTRATEGYS", "策略市集", "策略生成器"]
    cleaned = []
    for s in lines:
        s = sanitize_text(s)
        if not s or len(s) < 2:
            continue
        if any(j in s for j in junk_contains):
            continue
        cleaned.append(s)

    cleaned = uniq_keep_order(cleaned)

    # 合併短碎片
    merged = []
    for s in cleaned:
        if not merged:
            merged.append(s)
            continue
        prev = merged[-1]
        if len(prev) < 12 and len(s) < 20 and not re.search(r"[。！？:：]$", prev):
            merged[-1] = sanitize_text(prev + s)
        else:
            merged.append(s)

    # 切段
    paras = []
    buf = []
    def flush():
        nonlocal buf
        t = sanitize_text(" ".join(buf))
        if t:
            paras.append(t)
        buf = []

    for s in merged:
        is_heading_like = bool(re.match(r"^([一二三四五六七八九十]、|[0-9]+[.)])", s)) or (len(s) <= 28 and "：" in s)
        if is_heading_like and buf:
            flush()

        buf.append(s)

        cur = " ".join(buf)
        if re.search(r"[。！？]$", s) or len(cur) > 260:
            flush()

    flush()
    return paras


def safe_source_url(url: str) -> str:
    # 避免 Word 自動當外部連結觸發更新提示
    return url.replace("https://", "hxxps://").replace("http://", "hxxp://")


def fetch_image_bytes(url: str, timeout: int = 30):
    try:
        r = requests.get(url, timeout=timeout, headers={"User-Agent": "Mozilla/5.0"})
        if r.status_code != 200:
            return None, ""
        ct = (r.headers.get("content-type") or "").lower()
        return r.content, ct
    except Exception:
        return None, ""


def try_convert_to_png(image_bytes: bytes):
    """
    盡量把圖片轉 PNG（最穩）
    - 若 PIL 不可用或轉換失敗，就回傳 None
    """
    try:
        from PIL import Image  # noqa
    except Exception:
        return None

    try:
        im = Image.open(BytesIO(image_bytes))
        # 取第一幀（gif）
        if getattr(im, "is_animated", False):
            im.seek(0)
        if im.mode in ("RGBA", "LA") or (im.mode == "P" and "transparency" in im.info):
            out = BytesIO()
            im.convert("RGBA").save(out, format="PNG")
            return out.getvalue()
        else:
            out = BytesIO()
            im.convert("RGB").save(out, format="PNG")
            return out.getvalue()
    except Exception:
        return None


JS_COLLECT_TEXT = r"""
() => {
  const out = [];
  const norm = (s) => (s || "").replace(/\u00a0/g, " ").replace(/\s+/g, " ").trim();

  const visible = (el) => {
    try {
      const cs = getComputedStyle(el);
      if (cs.display === "none" || cs.visibility === "hidden" || cs.opacity === "0") return false;
      const r = el.getBoundingClientRect();
      if (r.width <= 0 || r.height <= 0) return false;
      return true;
    } catch { return true; }
  };

  const collectFromRoot = (root) => {
    if (!root) return;

    const elems = root.querySelectorAll ? root.querySelectorAll("h1,h2,h3,h4,p,li,blockquote,pre,span,div") : [];
    for (const el of elems) {
      if (!visible(el)) continue;
      const tag = (el.tagName || "").toLowerCase();
      if (["nav","header","footer","aside","button","input","textarea","select"].includes(tag)) continue;

      const t = norm(el.innerText || el.textContent || "");
      if (t && t.length >= 2 && t.length <= 4000) out.push(t);
    }

    const all = root.querySelectorAll ? root.querySelectorAll("*") : [];
    for (const el of all) {
      if (el.shadowRoot) collectFromRoot(el.shadowRoot);
    }
  };

  collectFromRoot(document);

  const bodyText = norm(document.body?.innerText || "");
  if (bodyText) out.push(bodyText);

  return out;
}
"""

JS_COLLECT_IMAGES = r"""
() => {
  const urls = [];
  const push = (u) => { if (u) urls.push(u); };

  const getBgUrls = (el) => {
    try {
      const cs = getComputedStyle(el);
      const bg = cs.backgroundImage || "";
      const matches = [...bg.matchAll(/url\(["']?(.*?)["']?\)/g)];
      return matches.map(m => m[1]).filter(Boolean);
    } catch { return []; }
  };

  const walk = (root) => {
    if (!root) return;
    if (root.shadowRoot) walk(root.shadowRoot);

    const nodes = root.querySelectorAll ? root.querySelectorAll("*") : [];
    for (const el of nodes) {
      const tag = (el.tagName || "").toLowerCase();
      if (tag === "img") {
        push(el.currentSrc || el.getAttribute("src") || el.getAttribute("data-src") || el.getAttribute("data-original") || el.getAttribute("data-lazy-src"));
      }
      for (const u of getBgUrls(el)) push(u);
      if (el.shadowRoot) walk(el);
    }
  };

  walk(document);
  return urls;
}
"""


async def wait_for_text_stable(page, interval_ms=700, stable_times=4, timeout_ms=45000):
    start = time.time()
    last_len = -1
    stable = 0
    while (time.time() - start) * 1000 < timeout_ms:
        try:
            length = await page.evaluate("() => (document.body?.innerText || '').length")
        except Exception:
            length = 0

        if length == last_len and length > 200:
            stable += 1
        else:
            stable = 0
        last_len = length

        if stable >= stable_times:
            return last_len

        await page.wait_for_timeout(interval_ms)
    return last_len


async def main():
    url = input("請輸入文章網址：\n").strip()
    if not url:
        print("未輸入網址，結束。")
        return

    # 用持久化 profile（反爬/渲染更穩）
    user_data_dir = os.path.join(os.getcwd(), "pw_profile_mcsm")

    async with async_playwright() as p:
        context = await p.chromium.launch_persistent_context(
            user_data_dir=user_data_dir,
            headless=False,
            viewport={"width": 1400, "height": 900},
            locale="zh-TW",
            args=["--disable-blink-features=AutomationControlled"],
        )
        page = await context.new_page()

        try:
            print("[INFO] Loading...")
            await page.goto(url, wait_until="domcontentloaded", timeout=60000)

            final_len = await wait_for_text_stable(page)
            print(f"[INFO] body_text_len(stable)={final_len}")

            # 觸發 lazy
            try:
                await page.evaluate("() => window.scrollTo(0, document.body.scrollHeight)")
                await page.wait_for_timeout(1200)
                await page.evaluate("() => window.scrollTo(0, 0)")
                await page.wait_for_timeout(600)
            except Exception:
                pass

            title = sanitize_text(await page.title()) or "Untitled"

            # 抓文字（含 shadow）
            lines = await page.evaluate(JS_COLLECT_TEXT)
            paras = normalize_paragraphs(lines)

            # 抓圖片（含 background-image）
            raw_imgs = await page.evaluate(JS_COLLECT_IMAGES)
            raw_imgs = uniq_keep_order(raw_imgs)

            img_urls = []
            for u in raw_imgs:
                try:
                    img_urls.append(urljoin(url, u))
                except Exception:
                    pass
            img_urls = uniq_keep_order(img_urls)

            print(f"[INFO] paras={len(paras)}, imgs={len(img_urls)}")

            if len(paras) < 3:
                print("[ERR] 正文仍不足（可能是 Canvas 文字）。若你要，我再給 OCR 版。")
                return

            # --- 產 docx（用 python-docx，Word 最穩） ---
            doc = Document()
            doc.add_heading(title, level=1)

            # 來源用 hxxps 避免 Word 當外部連結
            doc.add_paragraph(f"來源：{safe_source_url(url)}")
            doc.add_paragraph("")

            for ptxt in paras:
                ptxt = sanitize_text(ptxt)
                if ptxt:
                    doc.add_paragraph(ptxt)

            # 圖片（盡量轉 png 內嵌）
            # 若 PIL 不可用，則只內嵌 jpg/png，其他略過（避免壞檔/警告）
            added_any = False
            for u in img_urls:
                b, ct = fetch_image_bytes(u)
                if not b:
                    continue

                lower = u.lower()
                ct = (ct or "").lower()

                # svg 跳過（最容易出事）
                if "image/svg" in ct or lower.endswith(".svg"):
                    continue

                png = try_convert_to_png(b)
                if png is not None:
                    try:
                        doc.add_picture(BytesIO(png), width=Inches(6.0))
                        added_any = True
                    except Exception:
                        pass
                    continue

                # 沒 PIL 或轉失敗：只接受 jpg/png
                is_png = ("image/png" in ct) or lower.endswith(".png")
                is_jpg = ("image/jpeg" in ct) or ("image/jpg" in ct) or re.search(r"\.jpe?g(\?|#|$)", lower)
                if is_png or is_jpg:
                    try:
                        doc.add_picture(BytesIO(b), width=Inches(6.0))
                        added_any = True
                    except Exception:
                        pass

            # 檔名：用標題，避免覆蓋
            out_name = safe_filename(title) + ".docx"
            if os.path.exists(out_name):
                out_name = safe_filename(title) + "_" + datetime.now().strftime("%Y%m%d_%H%M%S") + ".docx"

            doc.save(out_name)
            print(f"[OK] 已輸出：{out_name}")
            if not added_any:
                print("[INFO] 圖片可能因格式（webp 等）或缺 PIL 而略過；文字已完成。")

        finally:
            await context.close()


if __name__ == "__main__":
    asyncio.run(main())

File: tools/grab_mcsm_docx.mjs
import fs from "fs";
import path from "path";
import readline from "readline";
import { chromium } from "playwright";
import { Document, Packer, Paragraph, HeadingLevel, TextRun, ImageRun } from "docx";
import JSZip from "jszip";
import { XMLParser, XMLBuilder } from "fast-xml-parser";

function ask(q) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => rl.question(q, (ans) => { rl.close(); resolve((ans || "").trim()); }));
}

function safeFilename(s, maxLen = 120) {
  s = (s || "").trim();
  if (!s) return "output";
  s = s.replace(/[<>:"/\\|?*\x00-\x1F]/g, "_");
  s = s.replace(/\s+/g, " ").trim();
  if (s.length > maxLen) s = s.slice(0, maxLen).trim();
  return s || "output";
}

function tsName() {
  const d = new Date();
  const pad = (n) => String(n).padStart(2, "0");
  return `${d.getFullYear()}${pad(d.getMonth() + 1)}${pad(d.getDate())}_${pad(d.getHours())}${pad(d.getMinutes())}${pad(d.getSeconds())}`;
}

function sanitizeXmlText(s) {
  s = (s || "").toString();
  let out = "";
  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i);
    if (code === 0x09 || code === 0x0A || code === 0x0D) { out += s[i]; continue; }
    if (code < 0x20) continue;
    if (code >= 0xD800 && code <= 0xDFFF) continue;
    if (code === 0xFFFE || code === 0xFFFF) continue;
    out += s[i];
  }
  out = out.replace(/\u00a0/g, " ").replace(/\s+/g, " ").trim();
  return out;
}

async function loadSharp() {
  try {
    const mod = await import("sharp");
    return mod.default || mod;
  } catch {
    return null;
  }
}

async function fetchBin(url) {
  const res = await fetch(url, { redirect: "follow" });
  if (!res.ok) return { buf: null, contentType: "" };
  const contentType = (res.headers.get("content-type") || "").toLowerCase();
  const ab = await res.arrayBuffer();
  return { buf: Buffer.from(ab), contentType };
}

function uniqKeepOrder(arr) {
  const seen = new Set();
  const out = [];
  for (const x of arr) {
    const k = (x || "").trim();
    if (!k) continue;
    if (seen.has(k)) continue;
    seen.add(k);
    out.push(k);
  }
  return out;
}

function normalizeTextLines(lines) {
  const junkContains = ["MULTISTRATEGYS", "策略市集", "策略生成器"];
  const filtered = lines
    .map(sanitizeXmlText)
    .filter(Boolean)
    .filter(s => s.length >= 2)
    .filter(s => !junkContains.some(j => s.includes(j)));

  const merged = [];
  for (const s of filtered) {
    if (!merged.length) { merged.push(s); continue; }
    const prev = merged[merged.length - 1];
    if (prev.length < 12 && s.length < 20 && !/[。！？:：]$/.test(prev)) {
      merged[merged.length - 1] = sanitizeXmlText(`${prev}${s}`);
    } else merged.push(s);
  }

  const paras = [];
  let buf = [];
  const flush = () => {
    const t = sanitizeXmlText(buf.join(" "));
    if (t) paras.push(t);
    buf = [];
  };

  for (const s of merged) {
    const isHeadingLike =
      /^([一二三四五六七八九十]、|[0-9]+[.)])/.test(s) ||
      (s.length <= 28 && /：/.test(s));

    if (isHeadingLike && buf.length) flush();

    buf.push(s);

    const cur = buf.join(" ");
    if (/[。！？]$/.test(s) || cur.length > 260) flush();
  }
  flush();

  return paras;
}

async function collectShadowText(page) {
  return await page.evaluate(() => {
    const out = [];
    const norm = (s) => (s || "").replace(/\u00a0/g, " ").replace(/\s+/g, " ").trim();

    const visible = (el) => {
      try {
        const cs = getComputedStyle(el);
        if (cs.display === "none" || cs.visibility === "hidden" || cs.opacity === "0") return false;
        const r = el.getBoundingClientRect();
        if (r.width <= 0 || r.height <= 0) return false;
        return true;
      } catch { return true; }
    };

    const collectFromRoot = (root) => {
      if (!root) return;
      const elems = root.querySelectorAll ? root.querySelectorAll("h1,h2,h3,h4,p,li,blockquote,pre,span,div") : [];
      for (const el of elems) {
        if (!visible(el)) continue;
        const tag = (el.tagName || "").toLowerCase();
        if (["nav","header","footer","aside","button","input","textarea","select"].includes(tag)) continue;
        const t = norm(el.innerText || el.textContent || "");
        if (t && t.length >= 2 && t.length <= 4000) out.push(t);
      }
      const all = root.querySelectorAll ? root.querySelectorAll("*") : [];
      for (const el of all) if (el.shadowRoot) collectFromRoot(el.shadowRoot);
    };

    collectFromRoot(document);

    const bodyText = norm(document.body?.innerText || "");
    if (bodyText) out.push(bodyText);

    return out;
  });
}

async function collectImages(page) {
  return await page.evaluate(() => {
    const urls = [];
    const push = (u) => { if (u) urls.push(u); };

    const getBgUrls = (el) => {
      try {
        const cs = getComputedStyle(el);
        const bg = cs.backgroundImage || "";
        const matches = [...bg.matchAll(/url\(["']?(.*?)["']?\)/g)];
        return matches.map(m => m[1]).filter(Boolean);
      } catch { return []; }
    };

    const walk = (root) => {
      if (!root) return;
      if (root.shadowRoot) walk(root.shadowRoot);

      const nodes = root.querySelectorAll ? root.querySelectorAll("*") : [];
      for (const el of nodes) {
        const tag = (el.tagName || "").toLowerCase();
        if (tag === "img") {
          push(el.currentSrc || el.getAttribute("src") || el.getAttribute("data-src") || el.getAttribute("data-original") || el.getAttribute("data-lazy-src"));
        }
        for (const u of getBgUrls(el)) push(u);
        if (el.shadowRoot) walk(el);
      }
    };

    walk(document);
    return urls;
  });
}

async function waitForTextStable(page, { intervalMs = 700, stableTimes = 4, timeoutMs = 45000 } = {}) {
  const start = Date.now();
  let lastLen = -1;
  let stable = 0;

  while (Date.now() - start < timeoutMs) {
    const len = await page.evaluate(() => (document.body?.innerText || "").length);
    if (len === lastLen && len > 200) stable++;
    else stable = 0;
    lastLen = len;
    if (stable >= stableTimes) return len;
    await page.waitForTimeout(intervalMs);
  }
  return lastLen;
}

async function getPngImageBuffer(url, sharp) {
  const { buf, contentType } = await fetchBin(url);
  if (!buf) return null;

  const ct = (contentType || "").toLowerCase();
  const lowerUrl = url.toLowerCase();

  // svg 直接跳過（最容易搞出 Word 問題）
  if (ct.includes("image/svg") || lowerUrl.endsWith(".svg")) return null;

  const isPng = ct.includes("image/png") || lowerUrl.endsWith(".png");
  const isJpg = ct.includes("image/jpeg") || ct.includes("image/jpg") || /\.jpe?g(\?|#|$)/i.test(url);

  if (!sharp) {
    // 沒 sharp 就只收 jpg/png（避免壞檔）
    if (isPng || isJpg) return buf;
    return null;
  }

  try {
    return await sharp(buf).png().toBuffer();
  } catch {
    return null;
  }
}

/**
 * ✅ 封包層清洗：刪掉所有 External relationships（Word 看到這個就會跳）
 */
async function stripExternalRelationships(docxBuffer) {
  const zip = await JSZip.loadAsync(docxBuffer);
  const parser = new XMLParser({ ignoreAttributes: false, attributeNamePrefix: "@_" });
  const builder = new XMLBuilder({ ignoreAttributes: false, attributeNamePrefix: "@_", suppressEmptyNode: true });

  // 只要在 word/_rels/*.rels 都掃
  const relPaths = Object.keys(zip.files).filter(p => /^word\/_rels\/.*\.rels$/i.test(p));
  for (const p of relPaths) {
    const xml = await zip.file(p).async("string");
    let obj;
    try { obj = parser.parse(xml); } catch { continue; }

    const rels = obj?.Relationships?.Relationship;
    if (!rels) continue;

    const arr = Array.isArray(rels) ? rels : [rels];

    const kept = arr.filter(r => {
      const tm = (r["@_TargetMode"] || "").toString();
      const target = (r["@_Target"] || "").toString();
      const t = target.toLowerCase();
      // 刪掉 external 或 任何看起來是網路/檔案路徑的 target
      if (tm.toLowerCase() === "external") return false;
      if (t.startsWith("http://") || t.startsWith("https://") || t.startsWith("file:")) return false;
      return true;
    });

    obj.Relationships.Relationship = kept.length === 1 ? kept[0] : kept;
    const newXml = builder.build(obj);
    zip.file(p, newXml);
  }

  return await zip.generateAsync({ type: "nodebuffer" });
}

async function main() {
  const url = await ask("請輸入文章網址：\n");
  if (!url) return console.log("未輸入網址，結束。");

  const userDataDir = path.join(process.cwd(), "pw_profile_mcsm");
  const context = await chromium.launchPersistentContext(userDataDir, {
    headless: false,
    viewport: { width: 1400, height: 900 },
    locale: "zh-TW",
    args: ["--disable-blink-features=AutomationControlled"],
  });

  const page = await context.newPage();

  try {
    console.log("[INFO] 打開瀏覽器中...");
    await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 });

    const finalLen = await waitForTextStable(page);
    console.log(`[INFO] body_text_len(stable)=${finalLen}`);

    await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
    await page.waitForTimeout(1200);
    await page.evaluate(() => window.scrollTo(0, 0));
    await page.waitForTimeout(600);

    const title = sanitizeXmlText((await page.title()) || "Untitled");

    const shadowLines = await collectShadowText(page);
    const mergedLines = uniqKeepOrder(shadowLines.map(sanitizeXmlText));
    const paras = normalizeTextLines(mergedLines);

    const rawImgs = await collectImages(page);
    const imgUrls = uniqKeepOrder(
      rawImgs.map(u => {
        try { return new URL(u, url).toString(); } catch { return ""; }
      }).filter(Boolean)
    );

    console.log(`[INFO] paras=${paras.length}, imgs=${imgUrls.length}`);

    if (paras.length < 3) {
      console.log("[ERR] 正文仍不足（可能是 Canvas 文字）。需要 OCR 才能保證。");
      return;
    }

    const sharp = await loadSharp();

    // ✅ 不建立超連結：來源只放純文字（而且把 https:// 變成 hxxps:// 避免任何自動判定）
    const safeUrlText = sanitizeXmlText(url.replace(/^https?:\/\//i, "hxxps://"));

    const children = [];
    children.push(new Paragraph({ text: title, heading: HeadingLevel.TITLE }));
    children.push(new Paragraph({ children: [new TextRun({ text: `來源：${safeUrlText}` })] }));
    children.push(new Paragraph({ text: "" }));

    for (const p0 of paras) {
      const p = sanitizeXmlText(p0);
      if (!p) continue;
      const isHeading =
        (p.length <= 28 && p.includes("：")) ||
        /^([一二三四五六七八九十]、|[0-9]+[.)])/.test(p);
      if (isHeading) children.push(new Paragraph({ text: p, heading: HeadingLevel.HEADING_2 }));
      else children.push(new Paragraph({ text: p }));
    }

    if (imgUrls.length) {
      children.push(new Paragraph({ text: "" }));
      children.push(new Paragraph({ text: "圖片", heading: HeadingLevel.HEADING_2 }));

      for (const u of imgUrls) {
        const pngBuf = await getPngImageBuffer(u, sharp);
        if (!pngBuf) continue;

        children.push(new Paragraph({
          children: [new ImageRun({ data: pngBuf, transformation: { width: 600, height: 380 } })],
        }));
      }
    }

    const doc = new Document({ sections: [{ children }] });
    let outBuf = await Packer.toBuffer(doc);

    // ✅ 最終保險：刪掉封包中所有 External relationships（Word 就不會跳）
    outBuf = await stripExternalRelationships(outBuf);

    const outName = `${safeFilename(title)}_${tsName()}.docx`;
    const outPath = path.join(process.cwd(), outName);
    fs.writeFileSync(outPath, outBuf);

    console.log(`[OK] 已輸出：${outPath}`);
  } catch (e) {
    console.error("[ERR]", e?.message || e);
  } finally {
    await context.close();
  }
}

main();
